File: ruby/004.rb
# A palindromic number reads the same both ways.
# The largest palindrome made from the product of
# two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product
# of two 3-digit numbers.

high = 999
low  = 100

def get_highest_palindrome high, low
  highest = 0
  high.downto low do |i|
    high.downto low do |j|
      sum = i * j
      if sum <= highest
        break
      end
      if is_palindrome(sum.to_s)
        highest = [highest, sum].max
      end
    end
  end
  return highest
end

def is_palindrome str
  return str == str.reverse
end

puts get_highest_palindrome(high, low)

File: go/src/euler/004.go
package main

import (
	"euler/lib"
	"fmt"
)

const max, min = 999, 100

func main() {
	highest := 0
	for i := max; i >= min; i-- {
		for j := max; j >= min; j-- {
			sum := i * j
			if sum < highest {
				break
			}
			if lib.IsPalindrome(sum) {
				highest = sum
				break
			}
		}
	}

	fmt.Println(highest)
}
